python/ucxx/examples/python_future_task.h

```diff
@@ -0,0 +1,159 @@
+
+ public:
+  ApplicationThread(PyObject* asyncioEventLoop,
+                    std::shared_ptr<std::mutex> incomingPoolMutex,
+                    FuturePoolPtr incomingPool,
+                    FuturePoolPtr readyPool)
+  {
+    ucxx_warn("Starting application thread");
+    _thread = std::thread(ApplicationThread::progressUntilSync,
+                          asyncioEventLoop,
+                          incomingPoolMutex,
+                          incomingPool,
+                          readyPool,
+                          std::ref(_stop));
+  }
+
+  ~ApplicationThread()
+  {
+    ucxx_warn("~ApplicationThread");
+    if (!_thread.joinable()) {
+      ucxx_warn("Application thread not running or already stopped");
+      return;
+    }
+
+    _stop = true;
+    _thread.join();
+  }
+
+  static void submit(std::shared_ptr<std::mutex> incomingPoolMutex,
+                     FuturePoolPtr incomingPool,
+                     FuturePoolPtr processingPool)
+  {
+    // ucxx_warn("Application submitting %lu tasks", incomingPool->size());
+    std::lock_guard<std::mutex> lock(*incomingPoolMutex);
+    for (auto it = incomingPool->begin(); it != incomingPool->end();) {
+      auto& task = *it;
+      processingPool->push_back(std::move(task));
+      it = incomingPool->erase(it);
+    }
+  }
+
+  static void processLoop(FuturePoolPtr processingPool, FuturePoolPtr readyPool)
+  {
+    // ucxx_warn("Processing %lu tasks", processingPool->size());
+    while (!processingPool->empty()) {
+      for (auto it = processingPool->begin(); it != processingPool->end();) {
+        auto& task   = *it;
+        auto& future = task.getFuture();
+
+        // 10 ms
+        std::future_status status = future.wait_for(std::chrono::duration<double>(0.01));
+        if (status == std::future_status::ready) {
+          ucxx_warn("Task %llu ready", future.get());
+          readyPool->push_back(std::move(task));
+          it = processingPool->erase(it);
+          continue;
+        }
+
+        ++it;
+      }
+    }
+  }
+
+  static void progressUntilSync(PyObject* asyncioEventLoop,
+                                std::shared_ptr<std::mutex> incomingPoolMutex,
+                                FuturePoolPtr incomingPool,
+                                FuturePoolPtr readyPool,
+                                const bool& stop)
+  {
+    ucxx_warn("Application thread started");
+    auto processingPool = std::make_shared<FuturePool>();
+    while (!stop) {
+      // ucxx_warn("Application thread loop");
+      ApplicationThread::submit(incomingPoolMutex, incomingPool, processingPool);
+      ApplicationThread::processLoop(processingPool, readyPool);
+    }
+  }
+};
+
+class Application {
+ private:
+  std::unique_ptr<ApplicationThread> _thread{nullptr};  ///< The progress thread object
+  std::shared_ptr<std::mutex> _incomingPoolMutex{
+    std::make_shared<std::mutex>()};  ///< Mutex to access the Python futures pool
```

Comment:
This is simply a `shared_ptr` because until now we've followed `cpplint`s default based on Google's styleguide, which [disallows non-const references](https://github.com/cpplint/cpplint/issues/148), which is what `std::mutex` would need to be to allow passing it to static member methods (e.g., `submit()`). I don't mind if we want to disable that and then rewrite this, but I'd prefer to do that (discuss/disable non-const references prohibition) in another issue or PR. Do you have strong opinions about it? Would you mind starting a discussion on it if so?

Reply:
Ah, ok. Nothing to see here then, I will collect my thoughts.